adrs/0001-cli-design.md

```diff
@@ -0,0 +1,306 @@
+  z_fcc     FCC plugin
+  z_osv     OSV plugin
+```
+
+selecting the **entities** command the user is presented with more sub commands
+
+```**commandline
+> griffon entities
+```
+and more context sensitive help:
+```commandline
+Usage: griffon entities [OPTIONS] COMMAND [ARGS]...
+
+  List and retrieve entities operations.
+
+Options:
+  --open-browser
+  --limit INTEGER
+  --help           Show this message and exit.
+
+Commands:
+  affects          https://osidb.prodsec.redhat.com/osidb/api/v1/affects
+  components       https://corgi-stage.prodsec.redhat.com/api/v1/components
+  flaws            https://osidb.prodsec.redhat.com/osidb/api/v1/flaws
+  product-streams  ...
+  trackers         https://osidb.prodsec.redhat.com/osidb/api/v1/trackers
+```
+
+and finally we are able to execute an operation, in this case returning all components with the name=component.
+```commandline
+> griffon entities components list --name curl 
+```
+
+Enabling [auto complete](https://click.palletsprojects.com/en/8.1.x/shell-completion/) lets the user easily discover 
+options and how to use commands/data.
+
+### Entity operations
+
+Each entity should minimally support:
+* list - list a set of entities
+* get - retrieving a specific entity 
+
+Some examples with components:
+
+```commandline
+griffon entities components list --name curl
+```
+
+```commandline
+griffon entities components get --uuid
+griffon entities components get --nvr
+griffon entities components get --purl
+```
+
+```commandline
+griffon entities flaws list 
+```
+
+```commandline
+griffon entities flaws get --cve_id 
+griffon entities flaws get --uuid 
+```
+
+### Service operations
+
+Modules for process/queries operations are explicitly isolated to enable reuse in other applications.
+
+We achieve isolation by enforcing interfaces (python Protocol) which constrain definition
+ ensuring the python modules can be easily used in future contexts.
+
+#### Query operations
+
+Perform read only cross service queries answering useful questions, some examples being:
+
+* Which unfixed CVE are affecting a component?
+* Which unfixed CVE are affecting a product + version + stream?
+* Given a CVE ID, what products are affected?
+* Given a CVE ID, what components are affected?
+* What products + version + stream contain a given component (e.g. full text search)?
+* What are the fixed CVEs for a product + version + stream?
+* What are the fixed CVEs for a component?
+* What are the won’t fix CVEs for a component?
+* What are the won’t fix CVEs for a product?
+* How many CVE’s are filed against a product + version
+
+For example, to get a list of cves affecting a specific component:
+```commandline
+> griffon queries component_cves --purl "pkg:rpm/redhat/curl@7.29.0-19.el7?arch=aarch64" --affectedness AFFECTED 
+ ```
+
+A set of 'canned' queries masks the complexity of calling up individual services and synthesizing results manually.
+
+#### Process operations
```

Comment:
This one (if I understand it correctly) might be interesting for OSIDB and process/workflow. For a real-life usage we would need to be able to eg. show a person's personal queue, team queue, unowned stuff affecting certain product etc. so it IR team can use the tool for processing the incoming reports/tickets...

Reply:
good point - we will need to define concept of user eg. krb creds gives us that in terms of user executing griffon locally ... eg. providing a process that mutates (CRUD) OSIDB will only be successful if krb creds give it to em ... I propose we go simple first and get CRUD in place before modelling higher order abstractions

Reply:
I agree with starting simple. CRUD is the base from our point of view. Workflow support is definitely something above. Having CRUD might help us bring some testers which could look at SFM2 queues but try to perform the actual work in OSIDB (through griffon) so we would have some early feedback.

Reply:
added GRIF-31